### DISCUSSION TITLE
ShouldUseConstructor is ignored at runtime

### DISCUSSION THREAD

**Discussion by Dresel** [LuckyPennySoftware/AutoMapper#3590] [Ideas]
Hi,

we have issues preventing AutoMapper to use private / internal constructors. Consider the following simplified example:

### Source/destination types

```csharp
public class Source
{
}

public class Destination
{
}
```

### Mapping configuration

```csharp
var config = new MapperConfiguration(cfg =>
{
    cfg.ShouldUseConstructor = c => false;
    cfg.CreateMap<Source, Destination>();
});
```

### Expected behavior

`mapper.Map<Destination>(new Source())` should throw an `AutoMapperConfigurationExpression` (No available constructor) exception.

### Actual behavior

`config.AssertConfigurationIsValid()` throws as expected. This is what got tested in the [unit tests](https://github.com/AutoMapper/AutoMapper/blob/master/src/UnitTests/ShouldUseConstructor.cs).

`mapper.Map<Destination>(new Source())` does still uses any ctor it finds.

### Steps to reproduce

```csharp
class Program
{
    static void Main(string[] args)
    {
        var config = new MapperConfiguration(cfg =>
        {
            cfg.ShouldUseConstructor = c => false;
            cfg.CreateMap<Source, Destination>();
        });

        //config.AssertConfigurationIsValid(); // This throws AutoMapperConfigurationExpression
        var destination = config.CreateMapper().Map<Destination>(new Source());
    }

    class Destination
    {
    }

    public class Source
    {
    }
}
```

**Comment by lbargaoanu** [comment]
That test shows the initial intent. I guess it's debatable.

**Reply by Dresel** [reply]
So ShouldUseConstructor is only meant for configuration validation but should not influence the mapping itself?

**Reply by Dresel** [reply]
Ok so I might have misinterpreted this a bit.

I just skimmed through the code - if I understood it correctly AutoMapper will always fallback to a parameterless or optional arguments only constructor, if no constructor mapping was found.

If there are potential constructor mappings, AutoMapper will choose one that satisfies ShouldUseConstructor.

Is this correct?

**Reply by Dresel** [reply]
Still odd that this would be reported as error though.

**Reply by lbargaoanu** [reply]
You should have a test that calls `AssertConfigurationIsValid`, making the difference irrelevant.

**Reply by Dresel** [reply]
I see. Wasn't aware that `AssertConfigurationIsValid` is considered mandatory. Thanks for the explanation and your time.

**Reply by lbargaoanu** [reply]
The initial discussion starts here https://github.com/AutoMapper/AutoMapper/pull/2698#issuecomment-399749750. I guess I see it the same way now, not a big deal. But if you feel strongly about it, @jbogard will probably accept a PR. Assuming the changes are trivial :)

**Reply by fiseni** [reply]
Hi,

In my opinion, the great benefit of `ShouldUseConstructor` was the ability to restrict the usage of given constructor types, more specifically the ability to exclude the private constructors. In this case, if no potential constructor mappings are found, it still will fall back to the private parameterless constructor (assuming it exists), which kind of defies the purpose 😞 What's the community consensus here? Is this behavior desirable?

Indeed, `AssertConfigurationIsValid` would catch this. But, even if by any chance you missed it, is it better to have a runtime exception rather than ending up with some inconsistent state?

Let me put a simple example

```c#
using AutoMapper;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Hosting;
using System;

namespace AutomapperSample
{
    public class Program
    {
        static void Main(string[] args)
        {
            var host = Host.CreateDefaultBuilder()
            .ConfigureServices((context, services) =>
            {
                services.AddScoped<MyApp>();
                services.AddAutoMapper(config =>
                {
                    config.ShouldUseConstructor = x => !x.IsPrivate;
                }, typeof(Program).Assembly);
            }).Build();

            using (var scope = host.Services.CreateScope())
            {
                var myApp = scope.ServiceProvider.GetRequiredService<MyApp>();

                myApp.Run();
            }
        }
    }

    public class MappingProfile : Profile
    {
        public MappingProfile()
        {
            ShouldUseConstructor = x => !x.IsPrivate;

            CreateMap<Bar, Foo>();
        }
    }

    public class MyApp
    {
        private readonly IMapper _mapper;

        public MyApp(IMapper mapper)
        {
            this._mapper = mapper;
        }

        public void Run()
        {
            var bar = new Bar { Name = "Test" };
            var foo = _mapper.Map<Foo>(bar);
        }
    }

    public class Bar
    {
        public string Name { get; set; }
    }

    public class Foo
    {
        public string Name { get; private set; }

        // Used specifically for EF, and should not be utilized for anything else.
        private Foo()
        {
            Console.WriteLine("The parameterless ctor is called");
        }

        // Just force the mapping to fail here
        public Foo(SomeComplexType someComplexType)
        {
            Console.WriteLine("The ctor with parameters is called");
        }
    }

    public class SomeComplexType { }
}
```

**Reply by lbargaoanu** [reply]
Not validating configuration is a capital offense :)

**Reply by fiseni** [reply]
Ok, I have no reply in that case 😄
